kernel/kernel-api/src/main/java/io/delta/kernel/internal/checkpoints/CheckpointInstance.java

```diff
@@ -36,19 +51,33 @@ public class CheckpointInstance
     public final long version;
     public final Optional<Integer> numParts;
 
+    public final CheckpointFormat format;
+
+    public final Optional<Path> filePath; // Guaranteed to be present for V2 checkpoints.
+
     public CheckpointInstance(String path) {
+        Preconditions.checkArgument(FileNames.isCheckpointFile(path),
+                "not a valid checkpoint file name");
+
         String[] pathParts = getPathName(path).split("\\.");
+        this.filePath = Optional.of(new Path(path));
 
-        if (pathParts.length == 3 && pathParts[1].equals("checkpoint") &&
-                pathParts[2].equals("parquet")) {
+        if (pathParts.length == 3 && pathParts[2].equals("parquet")) {
             // Classic checkpoint 00000000000000000010.checkpoint.parquet
             this.version = Long.parseLong(pathParts[0]);
             this.numParts = Optional.empty();
-        } else if (pathParts.length == 5 && pathParts[1].equals("checkpoint")
-                && pathParts[4].equals("parquet")) {
+            this.format = CheckpointFormat.CLASSIC;
+        } else if (pathParts.length == 5 && pathParts[4].equals("parquet")) {
             // Multi-part checkpoint 00000000000000000010.checkpoint.0000000001.0000000003.parquet
             this.version = Long.parseLong(pathParts[0]);
             this.numParts = Optional.of(Integer.parseInt(pathParts[3]));
+            this.format = CheckpointFormat.MULTI_PART;
+        } else if (pathParts.length == 4 && (pathParts[3].equals("parquet") ||
+                pathParts[3].equals("json"))) {
+            // V2 checkpoint 00000000000000000010.checkpoint.UUID.(parquet|json)
+            this.version = Long.parseLong(pathParts[0]);
+            this.numParts = Optional.empty();
+            this.format = CheckpointFormat.V2;
```

Comment:
move the above line `this.filePath = Optional.of(new Path(path));` here, so that the filePath is initialized just for the V2 checkpoint.

This makes the hashCode impl easy, you don't need if and else.